modules/images/webp-uploads/fallback.js

```diff
@@ -4,21 +4,49 @@ window.wpPerfLab = window.wpPerfLab || {};
 	window.wpPerfLab.webpUploadsFallbackWebpImages = function( media ) {
 		for ( var i = 0; i < media.length; i++ ) {
 			try {
-				if ( ! media[i].media_details.sources || ! media[i].media_details.sources['image/jpeg'] ) {
-					continue;
-				}
-
-				var ext = media[i].media_details.sources['image/jpeg'].file.match( /\.\w+$/i );
-				if ( ! ext || ! ext[0] ) {
-					continue;
-				}
 
-				var images = document.querySelectorAll( 'img.wp-image-' + media[i].id );
+				var images = document.querySelectorAll( 'img.wp-image-' + media[ i ].id ),
+					media_sources = media[ i ].media_details.sources;
 				for ( var j = 0; j < images.length; j++ ) {
-					images[j].src = images[j].src.replace( /\.webp$/i, ext[0] );
-					var srcset = images[j].getAttribute( 'srcset' );
+
+					if ( ! media_sources || ! media_sources['image/jpeg'] ) {
+						continue;
+					}
+
+					var srcset = images[ j ].getAttribute( 'srcset' ),
+						sizes = media[ i ].media_details.sizes,
+						sizes_keys = Object.keys( sizes ),
+						flag = true;
+
+					// If a full image is present in srcset, it should be updated.
+					if ( srcset && media_sources['image/webp'] ) {
+						srcset = srcset.replace( media_sources['image/webp'].file, media_sources['image/jpeg'].file );
+					}
+
+					for ( var k = 0; k < sizes_keys.length; k++ ) {
+						var media_sizes_sources = media[ i ].media_details.sizes[ sizes_keys[ k ] ].sources;
+						if ( ! media_sizes_sources || ! media_sizes_sources['image/webp'] || ! media_sizes_sources['image/jpeg'] ) {
+							continue;
+						}
+
+						// Check to see if the image src has any size set, then update it.
+						if ( flag && media_sizes_sources['image/webp'].source_url === images[ j ].src ) {
+							images[ j ].src = media_sizes_sources['image/jpeg'].source_url;
+							flag = false;
+						}
+
+						if ( srcset && media_sizes_sources['image/webp'] ) {
+							srcset = srcset.replace( media_sizes_sources['image/webp'].source_url, media_sizes_sources['image/jpeg'].source_url );
+						}
+					}
+
 					if ( srcset ) {
-						images[j].setAttribute( 'srcset', srcset.replace( /\.webp(\s)/ig, ext[0] + '$1' ) );
+						images[ j ].setAttribute( 'srcset', srcset );
+					}
+
+					// If the src has not been updated, then update the image src with the sources.
+					if ( flag && media_sources['image/webp'] ) {
+						images[ j ].src = images[ j ].src.replace( media_sources['image/webp'].file, media_sources['image/jpeg'].file );
 					}
```

Comment:
See above, based on my suggested change to the REST API response to be consistent with core, we will afterwards no longer need any special treatment for the `full` size, so this code can be removed. We also don't need the flag when we simply store the `src` in a variable.

You could then have something here exactly like you have right above for `srcset`:
```js
if ( src ) {
	images[ j ].setAttribute( 'src', src );
}
```

Reply:
Thank you. This feedback is addressed.